sys/include/net/sock/async/event.h

```diff
@@ -0,0 +1,257 @@
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * We then define an event handler in form of a function. The event handler
+ * checks if the triggering event was a receive event by checking the flags
+ * provided with sock_event_t::type. If it is a receive event it copies the
+ * incoming message to `buf` and its sender into `remote` using @ref
+ * sock_udp_recv(). Note, that we use @ref sock_udp_recv() non-blocking by
+ * setting `timeout` to 0. If an error occurs on receive, we just ignore it and
+ * return from the event handler.
+ *
+ * If we receive a message we use its `remote` to reply. In case of an error on
+ * send, we print an according message:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * int handler(sock_udp_t *sock, sock_async_flags_t type)
+ * {
+ *     if (type & SOCK_EVENT_MSG_RECV) {
+ *         sock_udp_ep_t remote;
+ *         ssize_t res;
+ *
+ *         if ((res = sock_udp_recv(sock, buf, sizeof(buf), 0,
+ *                                  &remote)) >= 0) {
+ *             puts("Received a message");
+ *             if (sock_udp_send(sock, buf, res, &remote) < 0) {
+ *                 puts("Error sending reply");
+ *             }
+ *         }
+ *     }
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * To be able to listen for incoming packets we bind the `sock` by setting a
+ * local end point with a port (`12345` in this case).
+ *
+ * We then proceed to create the `sock`. It is bound to `local` and thus listens
+ * for UDP packets with @ref udp_hdr_t::dst_port "destination port" `12345`.
+ * Since we don't need any further configuration we set the flags to 0.
+ * In case of an error we stop the program:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ *     sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+ *     sock_udp_t sock;
+ *
+ *     local.port = 12345;
+ *
+ *     if (sock_udp_create(&sock, &local, NULL, 0) < 0) {
+ *         puts("Error creating UDP sock");
+ *         return 1;
+ *     }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Finally, we initialize the event queue `queue`, initialize asynchronous event
+ * handling for `sock` using it and the previously defined event handler, and go
+ * into an endless loop to handle all occurring events on `queue`:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ *     event_queue_init(&queue);
+ *     sock_udp_event_init(&sock, &queue, handler);
+ *     event_loop(&queue);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief   Asynchronous sock using @ref sys_event definitions.
+ *
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
+#ifndef NET_SOCK_ASYNC_EVENT_H
+#define NET_SOCK_ASYNC_EVENT_H
+
+#include "event.h"
+/* guard required since `sock_dtls_types.h` might not be provided */
+#ifdef MODULE_SOCK_DTLS
+#include "net/sock/dtls.h"
+#endif  /* MODULE_SOCK_DTLS */
+#include "net/sock/ip.h"
+#include "net/sock/tcp.h"
+#include "net/sock/udp.h"
+#include "net/sock/async.h"
```

Comment:
When re-ordering this to a more sensible order I fall into the same trap as I am currently stuck in with `sock_dtls` in #12907 :-/

```diff
diff --git a/sys/include/net/sock/async/event.h b/sys/include/net/sock/async/event.h
index 9533334ff..02ad40cda 100644
--- a/sys/include/net/sock/async/event.h
+++ b/sys/include/net/sock/async/event.h
@@ -157,6 +157,7 @@
 #define NET_SOCK_ASYNC_EVENT_H
 
 #include "event.h"
+#include "net/sock/async.h"
 /* guard required since `sock_dtls_types.h` might not be provided */
 #ifdef MODULE_SOCK_DTLS
 #include "net/sock/dtls.h"
@@ -164,7 +165,6 @@
 #include "net/sock/ip.h"
 #include "net/sock/tcp.h"
 #include "net/sock/udp.h"
-#include "net/sock/async.h"
 
 #ifdef __cplusplus
 extern "C" {
```

```
In file included from /home/mlenders/Repositories/RIOT-OS/RIOT/sys/include/net/sock/ip.h:453,
                 from /home/mlenders/Repositories/RIOT-OS/RIOT/sys/include/net/sock/async.h:31,
                 from /home/mlenders/Repositories/RIOT-OS/RIOT/sys/include/net/sock/async/event.h:160,
                 from /home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/sock/async/event/sock_async_event.c:16:
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock/include/sock_types.h:58:36: error: unknown type name ‘sock_async_flags_t’
   58 |                                    sock_async_flags_t flags);
      |                                    ^~~~~~~~~~~~~~~~~~
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock/include/sock_types.h:81:9: error: unknown type name ‘gnrc_sock_reg_cb_t’
   81 |         gnrc_sock_reg_cb_t generic;     /**< generic version */
      |         ^~~~~~~~~~~~~~~~~~
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock/include/sock_types.h:83:9: error: unknown type name ‘sock_ip_cb_t’
   83 |         sock_ip_cb_t ip;                /**< IP version */
      |         ^~~~~~~~~~~~
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock/include/sock_types.h:86:9: error: unknown type name ‘sock_udp_cb_t’
   86 |         sock_udp_cb_t udp;              /**< UDP version */
      |         ^~~~~~~~~~~~~
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/pktdump
/home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock/include/sock_types.h:90:5: error: unknown type name ‘sock_async_ctx_t’
   90 |     sock_async_ctx_t async_ctx;         /**< asynchronous event context */
      |     ^~~~~~~~~~~~~~~~
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/net/gnrc/sock
```

Reply:
The problem is that there is a circular dependency, where-in both `sock_types.h`/`sock_dtls_types.h` of the network stack requires to include `net/sock/async.h` to add members related to asynchronous behavior to the type definition of the `sock`s, while `net/sock/async.h` needs to include `net/sock/<prot>.h` to access the `sock` types which in turn pulls in `sock_types.h`/`sock_dtls_types.h` for the actual type definition of the `sock`s. I'm currently unsure how to solve this :-/.